### PR TITLE
clang-format for discussion

### PR DESCRIPTION
**Pull request by rwcarlsen**
See #7289 for more discussion.  Github can't show the entire diff - you will have to fetch and checkout my branch to view the full diff locally.

### CODE REVIEW

**Comment by permcody** [comment]
I'm supportive of this change.

I will say that what I'd like to see is that we start in moose/framework and enforce it there for awhile. Then we should assist our application owners to work on applying the same rules to their individual modules. Ample notification should go out to users before we apply these diffs because it'll essentially wipe out any and all in-progress branches.

Tag @idaholab/moose-team for discussion.

**Comment by dschwen** [comment]
I suggest we do **not** make a PR like this but let moosebuild reformat the
files touched by a PR automatically (adding a small commit on top of the
PR). This would ensure that the formatting rules stay consistently enforced.

**Comment by rwcarlsen** [comment]
@dschwen - This is not meant to necessarily be the way we start using clang-format.  It is more intended as a discussion point for deciding if/how we want to tweak our style guide to enable usage of clang-format.

**Comment by rwcarlsen** [comment]
As far as enforcement methods go, I would prefer a CI test that checked for a non-empty diff between a PR and a clang-formatted version of a PR.

**Comment by YaqiWang** [comment]
I want this in place to save lot of reviews concerning the style. Can we limit the check for specific directories? I am not sure if we want to do this for a contrib folder.

**Comment by rwcarlsen** [comment]
Contrib would definitely be excluded from style enforcement.

**Comment by aeslaughter** [comment]
I think I like the auto-fix commit idea, if we know what is missing from
the format just fix it for people, if we can.

**Comment by rwcarlsen** [comment]
I'm only weakly opposed to an auto-fix approach.  Auto-fix breaks authorship and git blame for changed lines.  And I think auto-fix gradually for PRs as we convert to the new style would be super painful - peoples' in-flight changes to moose would continually be breaking over a period of several weeks.  Better to rip the bandaid off quickly and convert everything in one go - and then we can decide to auto-fix or CI check, etc. however we want after the full conversion.

**Comment by dschwen** [comment]
A ginormous fix everything PR would massively break authorship, too. I'd
prefer such a commit to be made by moosetest to make it obvious that it is
maintenance and not content.

**Comment by rwcarlsen** [comment]
Yes, but that is only a one-time break of authorship.  And we could have the ginormous PR have a ``moosetest`` author if we wanted.

**Comment by YaqiWang** [comment]
I personally like authors to fix their style, like what we did for trailing spaces. We can provide a script to help them fix the style. The blame page will not have the  `moosetest` author then.

**Comment by fdkong** [comment]
> Yes, but that is only a one-time break of authorship. And we could have the ginormous PR have a moosetest author if we wanted.

Is it technically possible NOT to break authorship when we do auto-fix formats?

**Comment by fdkong** [comment]
> I personally like authors to fix their style, like what we did for trailing spaces. We can provide a script to help them fix the style. The blame page will not have the moosetest author then.

I like this idea.

**Comment by friedmud** [comment]
I am against the idea of "auto-fix"... but I'm willing to be convinced.  Any "auto-fix" capability would definitely have to make sure that authorship is preserved (as well as possible anyway).

My main problem with it is that it will automatically generate conflicts if people are building on top of their PRs.  Let's say you are working on something and you have 3 commits.  You put those up as a PR because they work fine.  Then you continue to work in a new branch made from there and commit 2 more commits on top of those 3.

If those original 3 commits in the PR get modified when they go into MOOSE then when you do `git pull --rebase upstream devel` (or whatever your preferred rebase method is) you will get a shit-ton of conflicts... both with your original 3 commits _and_ possibly with your new 2 commits.

That's just nasty.

We haven't really had any trouble with the whitespace script... and just getting people to run that when they have whitespace issues.  If we make it easy to fix the issues then I think people can do it themselves.

I agree with @permcody that we should apply this to MOOSE first and see how it goes... then do modules.  There are a lot less in-flight branches that are modifying things in MOOSE than in the Modules.

**Comment by friedmud** [comment]
I need to amend my previous statement... if the formatting changes go in as a new commit on top of the original three... then there won't be any conflicts in those... but there still could be conflicts in the "new 2" commits you've made since you put up the PR.

File: framework/include/actions/Action.h [hunk 1]
```diff
@@ -30,7 +30,7 @@ class Executioner;
 class MooseApp;
 class Factory;
 
-template<>
+template <>
```
Comment (friedmud) [inline framework/include/actions/Action.h]: Can we turn this one off?  It seems unnecessary.

File: framework/include/actions/ActionWarehouse.h [hunk 2]
```diff
@@ -193,9 +219,9 @@ class ActionWarehouse :
    *
    * @param task The name of the task to find and build Actions for.
    */
-  void buildBuildableActions(const std::string &task);
+  void buildBuildableActions(const std::string & task);
 
-  std::vector<MooseSharedPointer<Action> > _all_ptrs;
+  std::vector<MooseSharedPointer<Action>> _all_ptrs;
```
Comment (friedmud) [inline framework/include/actions/ActionWarehouse.h]: Wow - I know we can do this now in C++11... but it still makes me uneasy :-)

File: framework/include/base/ExecuteMooseObjectWarehouse.h [hunk 3]
```diff
@@ -91,131 +96,123 @@ class ExecuteMooseObjectWarehouse : public MooseObjectWarehouse<T>
...
-    break;
-  case EXEC_NONLINEAR:
-    jacobianSetup(tid);
-    break;
-  case EXEC_LINEAR:
-    residualSetup(tid);
-    break;
-  default:
-    break;
+    case EXEC_INITIAL:
```
Comment (friedmud) [inline framework/include/base/ExecuteMooseObjectWarehouse.h]: This is fine with me.  We need to make sure to update our `.emacs` to do the right thing... ;-)

File: framework/include/base/FEProblem.h [hunk 4]
```diff
@@ -96,42 +96,47 @@ enum MooseNonlinearConvergenceReason
...
 enum MooseLinearConvergenceReason
 {
-  MOOSE_LINEAR_ITERATING                =  0,
+  MOOSE_LINEAR_ITERATING = 0,
   // MOOSE_CONVERGED_RTOL_NORMAL        =  1,
   // MOOSE_CONVERGED_ATOL_NORMAL        =  9,
-  MOOSE_CONVERGED_RTOL                  =  2,
-  MOOSE_CONVERGED_ATOL                  =  3,
-  MOOSE_CONVERGED_ITS                   =  4,
+  MOOSE_CONVERGED_RTOL = 2,
```
Comment (friedmud) [inline framework/include/base/FEProblem.h]: These are kind of a bummer... but it's ok if it enables automatic checking

File: framework/include/base/Factory.h [hunk 5]
```diff
@@ -29,77 +29,77 @@
...
-#define registerPartitioner(name)                   registerObject(name)
-
-#define registerNamedKernel(obj, name)              registerNamedObject(obj, name)
-#define registerNamedNodalKernel(obj, name)         registerNamedObject(obj, name)
-#define registerNamedBoundaryCondition(obj, name)   registerNamedObject(obj, name)
-#define registerNamedAux(obj, name)                 registerNamedObject(obj, name)
-#define registerNamedAuxKernel(name)                registerNamedObject(obj, name)
-#define registerNamedMaterial(obj, name)            registerNamedObject(obj, name)
-#define registerNamedPostprocessor(obj, name)       registerNamedObject(obj, name)
+#define registerKernel(name) registerObject(name)
```
Comment (friedmud) [inline framework/include/base/Factory.h]: This is a bummer too.  But understandable.

**Review by friedmud — COMMENTED** [review]
I think this is all acceptable.  I don't see anything in the formatting that makes me unhappy.

I would do a few things differently if I were manually formatting... but I think the tradeoff of having automatic checking is worth it...

File: framework/include/actions/Action.h (hunk 1 above)
Comment (permcody) [inline framework/include/actions/Action.h]: Agreed if it's possible... It probably isn't.

**Comment by dschwen** [comment]
I guess if we set up pre-commit hooks to do the reformatting it should be bearable. We might provide a helper script for users to do this. Otherwise I foresee a ton of failing pretests for trivial formatting reasons.

**Comment by permcody** [comment]
We can't assume that clang-format will be available or work properly on every machine. What if I just have intel for instance? I was chatting with Robert and we were thinking that CIVET should be extended so that a user can easily receive and apply the necessary patch for when they don't have these tools. I like the idea of a pre-commit hook though for most of our users.

**Comment by friedmud** [comment]
@dschwen Nearly every PR already "fails" our formatting... at least this will be automatic and won't require any intervention by a developer until the submitter runs a script and gets it right.

It's pretty normal for PRs to fail our whitespace script... but users are pretty good at fixing it.

**Comment by friedmud** [comment]
Nevermind my last comment - I see now.  You were talking about pre-commit.

**Comment by friedmud** [comment]
I see what @permcody is saying about not having clang on your machine.... hmmmm.

Getting a patch isn't a terrible idea... but is somewhat inconvenient.

I still think I could be convinced that "auto-fix" commits might be an ok idea.

Can we try that out in another repo to see how it would work?

**Comment by dschwen** [comment]
Not having clang is pretty irrelevant here. I was suggesting to provide a simple helper script that sets up the pre-commit hook if the user choses to do so. If you don't have clang it's your own problem - just don't setup the hook!

**Comment by dschwen** [comment]
The merge conflict issue is pretty nasty and for me kills the autofix idea :-(

**Comment by permcody** [comment]
It's not irrelevant from the point of view of coming up with a solution that will work for all users. Whatever we do, we'll need to design something that's easy for users to perform for that case where they don't have clang. Or worse, a potentially different version of clang that changes our format! I really hope that doesn't happen but who knows?

**Comment by friedmud** [comment]
@dschwen you're mixing up two things.

The issue with not having Clang is that there won't be any way for you to run the formatter to fix issues found by CIVET!

The whitespace script currently runs _everywhere_... so it's not a problem.

The suggestion by @permcody and @rwcarlsen about having a patch on CIVET that provides the fixes is a decent one... and might be the best way to go.  Let's just make sure we also add good instructions on how to apply it ;-)

**Comment by dschwen** [comment]
Yeah, applying a patch to get a PR to pass... that's adding a bit of a bar to new contributors. If we had a button to optionally have CIVET apply for it it would be a reasonable solution.

**Comment by friedmud** [comment]
I'm definitely ok with the idea of a button to have CIVET automatically
apply the patch... but I'm not sure exactly how that would work.  It
couldn't be applied directly to the branch that is in the PR... because
that's coming from a contributor's repo.

So... either CIVET would have to do a manual merge, push to devel and then
close the PR... or something...

**Comment by dschwen** [comment]
Submit a PR to the users repo?

**Comment by friedmud** [comment]
@dschwen That would be _really_ cool... and ultimately the best way to do things.

It should be possible.  What do you think @brianmoose ?

**Comment by brianmoose** [comment]
It is certainly possible to programmatically create a PR through the GitHub API. It might get a bit ugly since, in general, we don't have permission to create branches (or commits if we were talking about auto fixing PRs) on the contributors fork. So we would need a branch on idaholab/moose or some other repo to first apply the commit to.

**Comment by jwpeterson** [comment]
Wait, how does submitting a PR to the user's repo solve the self-conflict problem?  If they've moved on with development after opening their original PR, the PR CIVET gives back to them is going to conflict with that new development as well, right?

**Comment by dschwen** [comment]
It does not _solve_ the conflict problem, but it makes the fixing more apparent to the user. Ideally the user gets immediate feedback after opening the PR, can accept the fix and rebase its own branch _before_ moving on with development.

**Comment by rwcarlsen** [comment]
I think if CIVET just runs clang-format and prints the diff in CI output, that would be a good start.  Then inexperienced users would be able to look at the diff and see what needs to change.  More advanced users could copy/paste the diff and apply the patch using cli tools (or they could just use clang-format themselves).  We should definitely provide some pre-commit hooks that users can use locally for using clang-format too.

**Comment by friedmud** [comment]
I agree with @dschwen it should make fixing the problems immediate and transparent to the submitter (vs injecting a commit whenever the PR gets merged)

I'm fine with putting this in with CIVET just displaying a diff like @rwcarlsen recommends... but if CIVET could eventually be enhanced to put in a PR to the submitter that would be something I would personally make a lot of use of.

**Comment by permcody** [comment]
That may or may not work. New users are most likely to make really egregious formatting errors with code they edit. If a PR touches lines of code that are edited by other people a simple patch may not even apply at merge time.

Let's not over design this. Right now we have to tell people how to make all of these changes manually and for the most part people can handle that. There's no reason to change that work flow now. The only difference is that moosebuild gets to do that job through the diff we produce. We can have step by step instructions that tell you how to download and run the one command for applying the patch and squashing. It's something all users should understand anyway. I see a clickable link on CIVET with the patch and instructions right there that give the commands to apply it. All of this over-design is for the occasional guy that only has GCC (and no clang) on his system anyway. Everyone else will have clang-format and can just run a little script to fix their patch.

**Comment by friedmud** [comment]
That occasional guy with only GCC is _me_... like on Falcon.  I'm saying
that I would _personally_ enjoy getting a PR with the changes in it.

I agree that it's overdesigning (just telling people what to fix is
fine)... but it would be useful!

Let's go with the simple notification now and see how it goes.

Derek

**Comment by WilkAndy** [comment]
That occasional guy with only GCC is me too, and maybe lots of other people who mostly work on linux clusters.

You are all git masters, but most of the time i only half know what i'm doing.  Most people i interact with (who write the occasional MOOSE code) have some "git_help.txt" file they follow without understanding what's going on.  So it might be a good idea if the solution here doesn't involve much git (like applying patches and conflict resolution).

**Comment by dschwen** [comment]
There is also https://llvm.org/svn/llvm-project/cfe/trunk/tools/clang-format/git-clang-format for better git integration

**Comment by dschwen** [comment]
And there is https://commondatastorage.googleapis.com/chrome-infra-docs/flat/depot_tools/docs/html/git-hyper-blame.html to make the format change commit transparent w.r.t. authorship (unfortunately that does not help with the screwed up blame on the Github web interface).

**Comment by dschwen** [comment]
Can we please set 

```
AllowShortFunctionsOnASingleLine: true
```

We have a lot of small getter functions defined in headers that would get unnecessarily bulky with this set to false.

**Comment by dschwen** [comment]
And what about `SpaceAfterTemplateKeyword: false` ?! Adding this to the `.clang-format` file would get us closer to the status quo in style.

**Comment by friedmud** [comment]
I like both of those suggestions

**Comment by permcody** [comment]
I agree with Daniel and Derek, please make both of those changes.

**Comment by rwcarlsen** [comment]
As @dschwen pointed out, there are quite a few config options that crop up in each new clang release.  I tried to limit our usage of available options to those that work in 3.7.

Regarding AllowShortFunctionsOnASingleLine - this does some weird things w.r.t. constructor initializer lists without a column/line-length limit in place.  This is because clang has to use various heuristics to guess what can/should be put on a single line rather than matching against a specified line length.  This weirdness occurs despite our usage of the option ``ConstructorInitializerAllOnOneLineOrOnePerLine: true``.  Here is a sample of what I mean (pay attention to constructor initializer lists):

```c++
// unformatted/original file

class Foo {
public:
  Foo() : _bar(0), _baz(0) { }
  int _bar(0);
  double _baz(0);
};

class Foo {
public:
  Foo() :
    _bar(0), _baz(0) { }
  int _bar(0);
  double _baz(0);
};

class Foo {
public:
  Foo() : _bar(0),
    _baz(0) { }
  int _bar(0);
  double _baz(0);
};

class Foo {
public:
  Foo() : _bar(0), _baz(0)
  { }
  int _bar(0);
  double _baz(0);
};

```

and

```c++
// formatted
// ColumnLength: 0     # means no line length limit
// AllowShortFunctionsOnASingleLine: false
class Foo
{
public:
  Foo()
    : _bar(0), _baz(0)
  {
  }
  int _bar(0);
  double _baz(0);
};

class Foo
{
public:
  Foo()
    : _bar(0), _baz(0)
  {
  }
  int _bar(0);
  double _baz(0);
};

class Foo
{
public:
  Foo()
    : _bar(0),
      _baz(0)
  {
  }
  int _bar(0);
  double _baz(0);
};

class Foo
{
public:
  Foo()
    : _bar(0), _baz(0)
  {
  }
  int _bar(0);
  double _baz(0);
};
```

and

```c++
// formatted
// ColumnLength: 0     # means no line length limit
// AllowShortFunctionsOnASingleLine: true
class Foo
{
public:
  Foo() : _bar(0), _baz(0) {}
  int _bar(0);
  double _baz(0);
};

class Foo
{
public:
  Foo() : _bar(0), _baz(0) {}
  int _bar(0);
  double _baz(0);
};

class Foo
{
public:
  Foo() : _bar(0),
          _baz(0) {}
  int _bar(0);
  double _baz(0);
};

class Foo
{
public:
  Foo() : _bar(0), _baz(0)
  {
  }
  int _bar(0);
  double _baz(0);
};
```

and finally

```c++
// formatted
// ColumnLength: 80
// AllowShortFunctionsOnASingleLine: true
class Foo
{
public:
  Foo() : _bar(0), _baz(0) {}
  int _bar(0);
  double _baz(0);
};

class Foo
{
public:
  Foo() : _bar(0), _baz(0) {}
  int _bar(0);
  double _baz(0);
};

class Foo
{
public:
  Foo() : _bar(0), _baz(0) {}
  int _bar(0);
  double _baz(0);
};

class Foo
{
public:
  Foo() : _bar(0), _baz(0) {}
  int _bar(0);
  double _baz(0);
};
```

Note that in the last case clang always does the "right" thing - because we have a line length specified.  I think we should choose a column limit to put in our ``.clang-format``.

I also discovered the other day that you can turn formatting off for segments of code by doing this:

```c++
// clang-format off
std::map<std::string, std::string> foo = {
  "really really really long string key", "a key value",
  "short key"                          ", "another value",
  ... // etc.
};
// clang-format off
```

This should help alleviate concerns about choosing/using a line-length limit I think.  Thoughts?

**Comment by friedmud** [comment]
I agree - let's just choose a line limit.  I mean: does it do the right
thing if you set the line limit to 100,000?

But we can set a sensible limit anyway... ~200 or so would be ok with me.

**Comment by rwcarlsen** [comment]
@friedmud - the problem when you have options relating to allowing things to be on a single line it uses the line length to pack the lines smartly.  So we can't choose a limit longer than we actually want to see - otherwise Any function that fits within 200 character (for example) will be packed into a single line (regardless if the code writer put in line breaks before formatting).

**Comment by permcody** [comment]
With all of the "not quite right" things we are discovering, I wonder if we
should take a serious look at the tool @andrsd is using (patched astyle).
We might be able to distribute it with MOOSE.

**Comment by friedmud** [comment]
@rwcarlsen what do you mean by "any function"? Do you mean a function declaration and parameters... or the whole body too?

**Comment by rwcarlsen** [comment]
body too - that is what AllowShortFunctionsOnASingleLine does.

**Comment by andrsd** [comment]
> I wonder if we should take a serious look at the tool @andrsd is using (patched astyle).

I think I mentioned that it has its own problems and is not very good either. So, I would be against it.

Let's start another python project to parse C++ and reformat it. Anybody with CS degree? ;-)

**Comment by friedmud** [comment]
Bummer - too bad there aren't two different specifications for line lengths.  One for all code - and one for complete function bodies.

So, you're saying that if a function is 79 chars long and your line length limit is 80 then the whole function will get put on one line... no matter how many statements there are?

That really sucks and tends to make me think that we just want to turn off AllowShortFunctionsOnASingleLine.

I understand that will cause lots of diffs with what we're currently doing... but the alternative (getting whole functions reflowed onto one line) sucks too.

**Comment by dschwen** [comment]
I'm not buying it. This should mean single statement functions. You can also set it to `Inline` which would make this take effect only inside classes (which is the only place we want this anyways).

**Comment by friedmud** [comment]
If "Inline" works the way Daniel describes that would be ideal.  We do only
want these single line functions inside class headers.  They're actually
disallowed everywhere else.

Derek

**Comment by rwcarlsen** [comment]
A long column limit affects not just short functions - it affects *everything* that the format style generally believels "belongs" on a single line (e.g. a single statement, function headers, etc).  For example - with a 200 char column limit, this:

```c++
int foobarbaz(int areallylongvariablename,
    double anotherreallylongvariablename,
    double andanotherlongvariablename
    )
{
  int foo = 3;
}
```

becomes this:

```c++
int
foobarbaz(int areallylongvariablename, double anotherreallylongvariablename, double andanotherlongvariablename)
{
  int foo = 3;
}
```

**Comment by rwcarlsen** [comment]
ColumnLimit is effectively treated as a ColumnTarget by clang-format.

**Comment by YaqiWang** [comment]
Is this the rule? Only inline functions and functions can be in the source file should be in the header file.

**Comment by rwcarlsen** [comment]
I think we should just choose a reasonable column limit (to be enforced) and move forward with clang-format.  Anything between 80 and 100 chars is fine with me.  But keep in mind that clang-format will keep function calls, function headers, etc. packed to the limit - and will pack short one-line bodied functions that are under the limit onto a single line.  So we need to choose a length close to what we want to see most of the time.

**Comment by permcody** [comment]
I'm looking at some of my classes and I routinely flow past 100. I'd advocate an even bigger limit like 110 or 120. What do others think?

**Comment by dschwen** [comment]
I think 120 is a no go as long as clang-format collapses functions onto single lines if they fit in 120 chars.

**Comment by jwpeterson** [comment]
> I think 120 is a no go as long as clang-format collapses functions onto single lines if they fit in 120 chars.

Agreed, this is just a dumb feature IMO, and I don't understand why it can't just be turned off completely...

**Comment by permcody** [comment]
So the fact that we are stuck with picking some limit, do you guys have an opinion for a limit, or are we going to dump this clang format idea?

**Comment by rwcarlsen** [comment]
I prefer 80, but am okay with anything up to 100.  I frequently have three files open side-by-side and 120 would prevent me from doing that.  Many people don't have three giant monitors to spread our work over - especially external (including prospective) contributors.

**Comment by YaqiWang** [comment]
5000.

**Comment by permcody** [comment]
5000 doesn't work because clang-format will undo hard returns that you carefully put in there because the "can" fit within 5000 characters

**Comment by aeslaughter** [comment]
I am good with a hard limit if it allows us to have enforceable coding standards, maybe as clang-format goes along we can change it in the future. 

 I like 120, but don't care all that much.

**Comment by rwcarlsen** [comment]
@dschwen @jwpeterson @friedmud - preferred line lengths?

**Comment by dschwen** [comment]
Maybe clang-format is just not quite ready for us...

**Comment by jwpeterson** [comment]
> preferred line lengths?

My preferred *max* line length is probably around 120.  Unfortunately we are not talking about max line lengths, we're talking about making all lines as close to 120 characters long if possible, which is just dumb.

**Comment by YaqiWang** [comment]
oh, the formatter does not respect hard line breaks?! Will it do line breaks smarter by breaking places like after `,`?

**Comment by rwcarlsen** [comment]
Well - we can leave off the column limit if we are okay with a few strange quirks about things like what clang does with already weirdly formatted code in initializer lists.  I'd rather have no limit with the quirks than no clang-format.

**Comment by friedmud** [comment]
Can we get a look at what the codebase looks like without a limit?

**Comment by rwcarlsen** [comment]
This PR is roughly what it will look like - it was made with no column limit.

**Comment by friedmud** [comment]
Well - I don't like how it makes tiny one line functions into multi-line functions... but whatevs.  We'll survive.

I'm for going with this.  I just want consistent formatting and an automated way to do the formatting checking so that we don't spend so much time doing it for PRs.

**Comment by rwcarlsen** [comment]
If you put a tiny function on one line it will keep it on one line.  It just won't auto-update ones that are already broken onto multiple lines onto a single line (without the column limit).

**Comment by friedmud** [comment]
Then why is it spreading out so many tiny one line functions in this PR?

**Comment by rwcarlsen** [comment]
Because we have ``AllowShortFunctionsOnASingleLine: false`` in the ``.clang-format`` right now.  It was put there because that is our current moose style (per the guide).  But I would be happy to remove it (that's my preference) if others prefer it too.

**Comment by rwcarlsen** [comment]
This has served its purpose as a discussion area.  Closing.

**Pull request closed without merging**